rosetta/validator/requests.go

```diff
@@ -84,50 +98,102 @@ func (v *Validator) Request(request interface{}) error {
 	}
 
 	// Process validation errors we have found. Return the first one we encounter.
-	errs := err.(validator.ValidationErrors)
-	msg := errs[0].Tag()
+	// Validator library doesn't offer a sophisticated mechanism for passing detailed
+	// error information from the lower validation layers, so we use the `tag` field
+	// to identify what went wrong.
+	// In the case of blockchain/network misconfigurtion, we also use the `param` field
+	// to pass the acceptable value for the field.
 
-	switch msg {
+	verr := err.(validator.ValidationErrors)[0]
+
+	switch verr.Tag() {
 	case rosetta.BlockLength:
+		// Block hash has incorrect length.
+		blockHash, _ := verr.Value().(string)
 		return failure.InvalidBlockHash{
 			Description: failure.NewDescription(rosetta.BlockLength),
 			WantLength:  rosetta.HexIDSize,
+			HaveLength:  len(blockHash),
 		}
 
 	case rosetta.AddressLength:
+		// Account address has incorrect length.
+		address, _ := verr.Value().(string)
 		return failure.InvalidAccountAddress{
 			Description: failure.NewDescription(rosetta.AddressLength),
 			WantLength:  rosetta.HexAddressSize,
+			HaveLength:  len(address),
 		}
 
 	case rosetta.TxLength:
+		// Transaction hash has incorrect length.
+		txHash, _ := verr.Value().(string)
 		return failure.InvalidTransactionHash{
 			Description: failure.NewDescription(rosetta.TxLength),
 			WantLength:  rosetta.HexIDSize,
+			HaveLength:  len(txHash),
 		}
 
-	default:
-		return fmt.Errorf(msg)
-	}
-}
+	case networkFailTag:
+		// Network field is not referring to the network for which we are configured.
+		network, _ := verr.Value().(identifier.Network)
+		return failure.InvalidNetwork{
+			HaveNetwork:       network.Network,
+			AvailableNetworks: verr.Param(),
+			Description:       failure.NewDescription(unknownNetwork),
+		}
 
-func blockValidator(sl validator.StructLevel) {
-	rosBlockID := sl.Current().Interface().(identifier.Block)
-	if rosBlockID.Hash != "" && len(rosBlockID.Hash) != rosetta.HexIDSize {
-		sl.ReportError(rosBlockID.Hash, blockHashField, blockHashField, rosetta.BlockLength, "")
+	case blockchainFailTag:
+		// Blockchain field is not referring to the blockchain for which we are configured.
+		network, _ := verr.Value().(identifier.Network)
+		return failure.InvalidBlockchain{
+			HaveBlockchain:       network.Blockchain,
+			AvailableBlockchains: verr.Param(),
+			Description:          failure.NewDescription(unknownBlockchain),
+		}
+
+	default:
+		return fmt.Errorf(verr.Tag())
```

Comment:
```suggestion
		return errors.New(verr.Tag())
```

Since there's no formatting to be done, this is more efficient